### ISSUE TITLE
Bug: Adding observers that conform to both GroupObserverProtocol and ProcedureObserver to a GroupProcedure

### ISSUE THREAD

**Issue by swiftlyfalling** [ProcedureKit/ProcedureKit#574]
GroupProcedure has the following function to enable adding `GroupObserverProtocol` observers:

```swift
func add(observer: GroupObserverProtocol) {
    groupProtectedObservers.append(observer)
}
```

However, if adding an observer that conforms to **both** `GroupObserverProtocol` and `ProcedureObserver` (as in the case of ProcedureProfiler - see #573), this breaks the `ProcedureObserver` part of the observer.

`GroupProcedure.add(observer: GroupObserverProtocol)` gets called, and the observer only gets added to the groupProtectedObservers (and does not get added to the underlying Procedure observers).

**Comment by yageek** [comment]
Any suggestion to address this ?

**Comment by swiftlyfalling** [comment]
This may involve a bit of refactoring, so will wait on @danthorpe's input. (May be a short while.)

**Comment by danthorpe** [comment]
Hmm - will have a think. We kinda want to just call `super.add(observer: theObserver)` right? But the generics I guess make this a bit complicated? Is that the key issue here?

**Comment by yageek** [comment]
I tried to make a call to `super.add(observer: theObserver)` and yes, the generics did not help as `GroupObserver` is not related explicitly to `ProcedureObserver`.

**Comment by danthorpe** [comment]
Okay, I've been thinking about this, and I think the best solution is to combine the two protocols. Essentially, we have `ProcedureObserver` which already has:

```swift
func procedure(_ procedure: Procedure, didProduce newOperation: Operation)
```

and we have `GroupObserverProtocol` which has:

```swift
    func group(_ group: GroupProcedure, willAdd: Operation)

    func group(_ group: GroupProcedure, didAdd: Operation)
```

Given that `ProcedureObserver` is generic, as the `Procedure` argument is an associated type, when used from `GroupProcedure`, it is essentially exactly the same as `group(: didAdd:)`.

So, here's my proposal - get rid of `GroupObserverProtocol` entirely, and modify `ProcedureObserver` so that it has:

```swift
func procedure(_ procedure: Procedure, willAdd newOperation: Operation)
func procedure(_ procedure: Procedure, didAdd newOperation: Operation)
```

which `GroupProcedure` will use when new child operations are added, and `Procedure` will also use when the `produce(operation: Operation)` API is used - although I'm actually tempted to deprecate this API entirely.

Thoughts?

**Comment by danthorpe** [comment]
This is fixed by #583
